Option added to disable regular expression constraints.

SetTrend opened the pull request:
Currently, `EFCore.CheckConstraints` is missing an option to ignore regular expressions.

Regular expressions add costs to SQL execution and burdon to administering SQL Server instances.

It is beneficial to restrict checking regular expressions in the business layer or front-end only.

This pull requests adds a Boolean `UseRegex` option to `ValidationCheckConstraintOptions`. If `true`, regular expression constraints will not be generated.
<br/>

Fixes #18.

See PR #23 for a merged branch containing this pull request and #21.

roji commented on EFCore.CheckConstraints/CheckConstraintsExtensions.cs:
```diff
@@ -52,25 +93,115 @@ public static class CheckConstraintsExtensions
-                .UseValidationCheckConstraints();
-
+                .UseValidationCheckConstraints(validationCheckConstraintsOptionsAction);
+
```
Nit: please remove extra empty lines, leaving only one

SetTrend replied:
I'll gladly do ...

I did this because my intention was to visually separate the generic overloads group from the implementation group.

Just to make sure: I used **3** blank lines to separate them. Would **2** lines be alright for you? Or do you want me to wipe the separation completely and use just **1** blank line?

roji replied:
I'd rather we followed the existing styling (which also corresponds to EF Core itself) of just 1 blank line.

SetTrend replied:
Done.

roji commented on EFCore.CheckConstraints/EFCore.CheckConstraints.csproj:
```diff
@@ -19,6 +19,10 @@
     <RepositoryUrl>git://github.com/efcore/EFCore.CheckConstraints</RepositoryUrl>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(Platform)'=='AnyCPU'">
```
This seems like an accidental addition that should be removed.

SetTrend replied:
That group of lines in the project enables the XML documentation comments:

```xml
<PropertyGroup Condition="'$(Platform)'=='AnyCPU'">
  <DocumentationFile>obj\EFCore.CheckConstraints.xml</DocumentationFile>
</PropertyGroup>
```

roji replied:
Please add `<GenerateDocumentationFile>true</GenerateDocumentationFile>` instead in the global properties section of the csproj, [as documented here](https://docs.microsoft.com/en-us/dotnet/csharp/codedoc).

SetTrend replied:
Done. (Commit postponed until comments have all been resolved.)

Looking forward to your suggestions on my reply to your other comment.

roji commented on EFCore.CheckConstraints/Internal/CheckConstraintsConventionSetPlugin.cs:
```diff
@@ -1,18 +1,37 @@
 
 namespace EFCore.CheckConstraints.Internal
 {
+    /// <summary>
```
Please remove these comments - we don't document APIs in Internal namespaces. Same in the other files.

SetTrend replied:
Would you like me to remove them all?

I added XML documentation comments, so the user will get IntelliSense when using _CheckConstraints_:

![IntelliSense 1](https://user-images.githubusercontent.com/9283914/110150790-f23a0600-7ddf-11eb-8305-fef898f7c430.png)

![IntelliSense 2](https://user-images.githubusercontent.com/9283914/110150796-f403c980-7ddf-11eb-9ea4-b06fb496ec66.png)
<br/>

In order to avoid numerous warnings, I commented all the code until no more warnings occured.

How do you like me to proceed?

roji replied:
Please remove all XML docs from types in Internal namespaces - this code isn't meant to be used by users and is very similar to actual `internal` (or `private`) code.

SetTrend replied:
`ValidationCheckConstraintOptionsBuilder` is public, though (see my screenshot above).

Shouldn't `ValidationCheckConstraintOptionsBuilder` then better be moved outside the `internal` folder?

roji replied:
You're right - this specific type indeed should be public... Can you please move it out of Internal?

roji replied:
But for other cases of types that really should be Internal, we should remove the docs.

roji commented on EFCore.CheckConstraints/Internal/CheckConstraintsOptionsExtension.cs:
```diff
@@ -2,21 +2,38 @@
         private bool _discriminatorCheckConstraintsEnabled;
         private ValidationCheckConstraintOptions _validationCheckConstraintsOptions;
 
+
```
nit: extra empty lines (here and below)

SetTrend replied:
Done.

roji commented on EFCore.CheckConstraints/Internal/ValidationCheckConstraintConvention.cs:
```diff
@@ -39,7 +72,7 @@ public class ValidationCheckConstraintConvention : IModelFinalizingConvention
             _intTypeMapping = relationalTypeMappingSource.FindMapping(typeof(int));
 
-            _supportsRegex = SupportsRegex;
+            _supportsRegex = options.UseRegex && SupportsRegex;
```
Let's rename from `_supportsRegex` (which is no longer about database support) to `_useRegex` (which should still combine `options.UseRegex` and `SupportsRegex` as you've done)

SetTrend replied:
done.

roji reviewed (changes requested):
Thanks for this PR, it makes sense to me - see comments below.

Also, it seems that UseRegex is off by default - is this intentional? It seems more appropriate to have this as an opt-out, rather than as an opt-in, no?

SetTrend commented:
> 
> 
> Thanks for this PR, it makes sense to me - see comments below.
> 
> Also, it seems that UseRegex is off by default - is this intentional? It seems more appropriate to have this as an opt-out, rather than as an opt-in, no?

D'oh ... Yes ... I made a mistake here.

I corrected it now (didn't commit yet, awaiting your replies to the other comments before checking in).

I also added a number of unit tests to verify the result of using the different `BuilderOptions`:

_Preview:_
![OptionsBuilderTest](https://user-images.githubusercontent.com/9283914/110157198-36c99f80-7de8-11eb-957b-6254781783b9.png)

roji reviewed (commented):
Thanks, this looks almost ready - see the remaining unresolved comments above.

SetTrend commented:
Here we go. Everything's done.

roji reviewed (approved):
Thanks!
